### ISSUE TITLE
Make EH debugging not event order dependent

### ISSUE THREAD

**Issue by isidorn** [Microsoft/vscode#29884]
Currently the way extension host debugging is implemented it is very much depended on event ordering and timing. We should invest in making this less fragile.

Issue where this is showing up https://github.com/Microsoft/vscode/issues/29334

**Comment by weinand** [comment]
The fundamental problem is that the extension host (EH) is just a regular node.js process that is unaware of the fact that it is used as an EH. So a termination event originating from the node.js process conveys no information whether it is a result of a crash or the first part of an EH restart sequence.

The VS Code code that manages the EH process knows that it first kills a EH process and then starts a new EH as part of a restart. But it is very difficult to associate this high-level knowledge (an event) with the low-level events coming from the node.js process.

The problem that lead to the issue at hand was that the order of the high-level event and low-level events was flipped because of some performance optimisation. So currently we use an algorithm that requires a fixed order between these events.

A better (but still fragile) approach would be this:
If a node.js termination event is received and there was no restart event some X ms before or X ms after, the termination event indicates a crash and the debug session terminates. Otherwise the termination event is part of a restart sequence and the debug session needs to be restarted.

The best approach to fix this issue would be to add additional context information to the low-level events sent out by the node.js runtime. With that VS Code would know that a terminated event is part of a restart sequence (and there would be no need to associate events with some heuristic).

**Comment by weinand** [comment]
Ideally the logic would be this:

- the debug UI receives a reattach event (from Ben) whenever the EH was restarted (e.g. when loading a workbench or reloading the window).
- this triggers the creation of a new "Process" (debug terminology) which tracks a new DA (that itself tracks the new node.js EH process). The debugger UI will immediately "take over" this Process. 
- the old node.js EH is still tracked by the old Process and DA but it should no longer be reflected in the UI (because the new EH is already shown). If node.js eventually terminates, the TerminatedEvent will make the debugger frontend cleanup the "Process" and DA. 

In this model two Processes (and their associated DAs and EHs) can exist in parallel but only one is reflected in the UI. As soon as the new EH is running, the old session disappears from the UI (even if the node.js EH process hasn't terminated).

If the EH crashes or terminates, it will send a TerminatedEvent through the DA and the debugger frontend will shutdown the session. Since no new "Process" was started, the UI still reflects the crashed EH and no "restart" will happen.

**Comment by weinand** [comment]
A simpler alternative is to avoid that EH process can run in parallel.
In this model launching of the new EH process would be delayed until the previous EH has completely terminated. This could be a first step towards the more robust solution from above.

**Comment by isidorn** [comment]
I have looked into this and in the end settled on a pragmatic solution which seems to work. Please keep in mind that both the start of EH debugging and reload trigger the attach event from the workbench thus they fall into the same camp.

* Reattach event now always creates a brand new "Process"
* If there is already a EH Process when a reattach happens that Process is disposed. Why we need to do this and not simply let DA handle that? Because in the case when we start EH debugging we have a Process which is not yet attached and in that case DA will not send any terminate signal since nothing died, so we need to get rid of this empty Process and start using the real thing.

I have tested this out and it works nicely for me (both realod and open folder and start of eh debugging).
Since I am out for the next 10 days in case something seems broken just revert my commit attached to this issue.

This solution compared to the ideal logic from two posts above does not have the problem of having to hide a process in the ui (which we currently do not support).

Try it out and let me know how it goes.

**Comment by weinand** [comment]
@isidorn sounds good, I'll give it a thorough try...

**Comment by weinand** [comment]
"Reload" Window and "Open..." works great on all three platforms.

But "Restart" is broken now.

Since I deem the former actions more important than the latter, I will not revert the commit but create a separate issue for the "Restart" bug.
